File: 0077-combinations/0077-combinations.cpp
class Solution {
public:
    void back(vector<vector<int>>&ans,vector<int>curr,int i,int n,int k){
        if(i>n+1)return;
        if(curr.size()==k){
            ans.push_back(curr);
            return;
        }
        for(int j=i;j<=n;j++){
            curr.push_back(j);
            back(ans,curr,j+1,n,k);
            curr.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>>ans;
        vector<int>curr;
        back(ans,curr,1,n,k);
        return ans;
    }
};

File: 0077-combinations/0077-combinations.py
# def backtrack(curr, first_num):
#     if len(curr) == k:
#         ans.append(curr[:])
#         return

    
#     for num in range(first_num, n + 1):
#         curr.append(num)
#         backtrack(curr, num + 1)
#         curr.pop()


# class Solution:
#     def combine(self, n: int, k: int) -> List[List[int]]:
#         def backtrack(curr, first_num):
#             if len(curr) == k:
#                 ans.append(curr[:])
#                 return

#             need = k - len(curr)
#             remain = n - first_num + 1
#             available = remain - need
            
#             for num in range(first_num, first_num + available + 1):
#                 curr.append(num)
#                 backtrack(curr, num + 1)
#                 curr.pop()

#             return
        
#         ans = []
#         backtrack([], 1)
#         return ans

class Solution:
    def combine(self, n, k):
        if k == 0:
            return [[]]
        return [pre + [i] for i in range(k, n+1) for pre in self.combine(i-1, k-1)]